src/storage-controller/src/lib.rs

```diff
@@ -1643,26 +1668,68 @@ where
                 }
             }
 
-            // Check if the collection is for a Webhook source, unregister if so.
+            // Check if a collection is managed by the storage-controller itself.
             let collection = self.collections.get(&id);
             if let Some(CollectionState { description, .. }) = collection {
-                if description.data_source == DataSource::Webhook && frontier.is_empty() {
-                    // Unregister our collection from the manager so writes should no longer occur.
-                    self.collection_manager.unregister_collection(id).await;
+                let is_controller_managed = matches!(
+                    description.data_source,
+                    DataSource::Webhook
+                        | DataSource::Introspection(_)
+                        | DataSource::Other(DataSourceOther::TableWrites)
+                );
 
-                    pending_source_drops.push(id);
-                    // Normally `clusterd` will emit this StorageResponse when it knows we can
-                    // drop an ID, but since Webhook sources don't run on a cluster, we manually
-                    // emit this event here.
-                    let _ = self
-                        .internal_response_sender
-                        .send(StorageResponse::DroppedIds([id].into()));
+                // Normally `clusterd` will emit this StorageResponse when it knows we can drop an
+                // ID, but some collections are managed by the storage controller itself, so we
+                // must emit this event here.
+                if is_controller_managed && read_frontier.is_empty() {
+                    pending_collection_drops.push(id);
```

Comment:
I believe `pending_collection_drops` and `pending_source_drops` intersect, which will create negative multiplicities in [the new code that removes shard mappings](https://github.com/MaterializeInc/materialize/pull/25173/files#diff-3d40535ccc11f976645c3c15ed4b6b1a08b7c04de0d21085ba2a4a891b6e5422R1749-R1753).

Reply:
Ahhh good call, updated to de-dupe this list before we append shard mappings!